# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/jsapi/TransceiverImpl.h
# Commit: 970ed33359b5
# Full Hash: 970ed33359b5223277984f77c521faaa0079aa88
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 2
# Description:
#   Bug 1654112 - Implement MediaConduitControl for controlling conduits async through state mirroring. r=ng,bwc
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D112868
# ==============================================================================

diff -r ed2e505b5873 -r 970ed33359b5 dom/media/webrtc/jsapi/TransceiverImpl.h
--- a/dom/media/webrtc/jsapi/TransceiverImpl.h	Wed Apr 21 01:13:43 2021 +0200
+++ b/dom/media/webrtc/jsapi/TransceiverImpl.h	Tue Aug 10 14:13:49 2021 +0200
@@ -5,6 +5,7 @@
 #define _TRANSCEIVERIMPL_H_
 
 #include <string>
+#include "libwebrtcglue/MediaConduitControl.h"
 #include "mozilla/RefPtr.h"
 #include "nsCOMPtr.h"
 #include "nsISerialEventTarget.h"
@@ -19,7 +20,6 @@
 namespace mozilla {
 class PeerIdentity;
 class JsepTransceiver;
-enum class MediaSessionConduitLocalDirection : int;
 class MediaSessionConduit;
 class VideoSessionConduit;
 class AudioSessionConduit;
@@ -141,24 +141,45 @@
       const JsepTrackNegotiatedDetails& aDetails,
       std::vector<VideoCodecConfig>* aConfigs);
 
-  static void UpdateConduitRtpExtmap(
-      MediaSessionConduit& aConduit, const JsepTrackNegotiatedDetails& aDetails,
-      const MediaSessionConduitLocalDirection aDir);
+  AbstractCanonical<bool>* CanonicalReceiving() { return &mReceiving; }
+  AbstractCanonical<bool>* CanonicalTransmitting() { return &mTransmitting; }
+  AbstractCanonical<Ssrcs>* CanonicalLocalSsrcs() { return &mLocalSsrcs; }
+  AbstractCanonical<std::string>* CanonicalLocalCname() { return &mLocalCname; }
+  AbstractCanonical<std::string>* CanonicalLocalMid() { return &mLocalMid; }
+  AbstractCanonical<std::string>* CanonicalSyncGroup() { return &mSyncGroup; }
+  AbstractCanonical<RtpExtList>* CanonicalLocalSendRtpExtensions() {
+    return &mLocalSendRtpExtensions;
+  }
+  AbstractCanonical<Maybe<AudioCodecConfig>>* CanonicalAudioSendCodec() {
+    return &mAudioSendCodec;
+  }
+  AbstractCanonical<Ssrcs>* CanonicalLocalVideoRtxSsrcs() {
+    return &mLocalVideoRtxSsrcs;
+  }
+  AbstractCanonical<Maybe<VideoCodecConfig>>* CanonicalVideoSendCodec() {
+    return &mVideoSendCodec;
+  }
+  AbstractCanonical<Maybe<RtpRtcpConfig>>* CanonicalVideoSendRtpRtcpConfig() {
+    return &mVideoSendRtpRtcpConfig;
+  }
+  AbstractCanonical<webrtc::VideoCodecMode>* CanonicalVideoCodecMode() {
+    return &mVideoCodecMode;
+  }
 
  private:
   virtual ~TransceiverImpl();
   void InitAudio();
   void InitVideo();
+  void InitConduitControl();
   nsresult UpdateAudioConduit();
   nsresult UpdateVideoConduit();
-  nsresult ConfigureVideoCodecMode(VideoSessionConduit& aConduit);
+  nsresult ConfigureVideoCodecMode();
   void Stop();
 
   nsCOMPtr<nsPIDOMWindowInner> mWindow;
   const std::string mPCHandle;
   RefPtr<MediaTransportHandler> mTransportHandler;
   RefPtr<JsepTransceiver> mJsepTransceiver;
-  std::string mMid;
   bool mHaveSetupTransport;
   nsCOMPtr<nsISerialEventTarget> mMainThread;
   nsCOMPtr<nsISerialEventTarget> mStsThread;
@@ -166,6 +187,8 @@
   // state for webrtc.org that is shared between all transceivers
   RefPtr<WebrtcCallWrapper> mCallWrapper;
   RefPtr<MediaSessionConduit> mConduit;
+  // Call thread only.
+  RefPtr<MediaConduitController> mConduitController;
   RefPtr<MediaPipelineTransmit> mTransmitPipeline;
   // The spec says both RTCRtpReceiver and RTCRtpSender have a slot for
   // an RTCDtlsTransport.  They are always the same, so we'll store it
@@ -178,6 +201,21 @@
   RefPtr<dom::RTCRtpReceiver> mReceiver;
   // TODO(bug 1616937): Move this to RTCRtpSender
   RefPtr<dom::RTCDTMFSender> mDtmf;
+
+  Canonical<bool> mReceiving;
+  Canonical<bool> mTransmitting;
+  Canonical<Ssrcs> mLocalSsrcs;
+  Canonical<Ssrcs> mLocalVideoRtxSsrcs;
+  Canonical<std::string> mLocalCname;
+  Canonical<std::string> mLocalMid;
+  Canonical<std::string> mSyncGroup;
+  Canonical<RtpExtList> mLocalSendRtpExtensions;
+
+  Canonical<Maybe<AudioCodecConfig>> mAudioSendCodec;
+
+  Canonical<Maybe<VideoCodecConfig>> mVideoSendCodec;
+  Canonical<Maybe<RtpRtcpConfig>> mVideoSendRtpRtcpConfig;
+  Canonical<webrtc::VideoCodecMode> mVideoCodecMode;
 };
 
 }  // namespace mozilla